File: app/src/main/java/com/juniperphoton/myerlist/widget/SelectCategoryView.java
package com.juniperphoton.myerlist.widget;

import android.content.Context;
import android.support.v4.content.ContextCompat;
import android.util.AttributeSet;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;

import com.juniperphoton.myerlist.R;
import com.juniperphoton.myerlist.model.ToDoCategory;
import com.juniperphoton.myerlist.realm.RealmUtils;
import com.juniperphoton.myerlist.util.DisplayUtil;

import io.realm.Realm;
import io.realm.RealmResults;

public class SelectCategoryView extends LinearLayout implements ViewGroup.OnClickListener {
    private static final String TAG = "SelectCategoryView";

    private Context mContext;
    private OnSelectionChangedCallback mCallback;
    private int mSelectedIndex = 0;

    public SelectCategoryView(Context context, AttributeSet attrs) {
        super(context, attrs);
        mContext = context;
    }

    public void setOnSelectionChangedCallback(OnSelectionChangedCallback callback) {
        mCallback = callback;
    }

    public int getSelectedIndex() {
        return mSelectedIndex;
    }

    public void invokeCallback() {
        if (mCallback != null) {
            mCallback.onSelectionChanged(mSelectedIndex);
        }
    }

    public void setSelected(int index) {
        mSelectedIndex = index;
        updateUi((CateCircleView) getChildAt(mSelectedIndex));
        invokeCallback();
    }

    public void toggleSelection(boolean plus) {
        Log.d(TAG, "toggleSelection:" + plus);
        int target = mSelectedIndex + (plus ? 1 : -1);
        if (target >= getChildCount()) target = 0;
        if (target < 0) target = getChildCount() - 1;
        mSelectedIndex = target;
        updateUi((CateCircleView) getChildAt(mSelectedIndex));
        invokeCallback();
    }

    public void makeViews() {
        removeAllViews();
        Realm realm = RealmUtils.getMainInstance();
        realm.beginTransaction();
        RealmResults<ToDoCategory> categories = realm.where(ToDoCategory.class).findAll();
        for (ToDoCategory toDoCategory : categories) {
            if (toDoCategory.getId() >= 0) {
                CateCircleView circleView = new CateCircleView(mContext, null);
                LinearLayout.LayoutParams layoutParams = new LinearLayout.LayoutParams(
                        DisplayUtil.getDimenInPixel(24, mContext),
                        DisplayUtil.getDimenInPixel(24, mContext));
                layoutParams.setMargins(DisplayUtil.getDimenInPixel(8, mContext), 0, 0, 0);
                circleView.setLayoutParams(layoutParams);
                circleView.setColor(toDoCategory.getIntColor());
                circleView.setOnClickListener(this);
                addView(circleView);
            }
        }
        CateCircleView circleView = new CateCircleView(mContext, null);
        LinearLayout.LayoutParams layoutParams = new LinearLayout.LayoutParams(
                DisplayUtil.getDimenInPixel(24, mContext),
                DisplayUtil.getDimenInPixel(24, mContext));
        circleView.setLayoutParams(layoutParams);
        circleView.setColor(ContextCompat.getColor(mContext, R.color.MyerListBlue));
        circleView.setOnClickListener(this);
        addView(circleView, 0);
        getChildAt(0).setSelected(true);
        realm.commitTransaction();
    }

    @Override
    public void onClick(View v) {
        updateUi((CateCircleView) v);
        invokeCallback();
    }

    private void updateUi(CateCircleView circleView) {
        for (int i = 0; i < getChildCount(); i++) {
            CateCircleView cateCircleView = (CateCircleView) getChildAt(i);
            if (cateCircleView == circleView) {
                cateCircleView.setSelected(true);
                mSelectedIndex = i;
            } else {
                cateCircleView.setSelected(false);
            }
        }
    }

    public interface OnSelectionChangedCallback {
        void onSelectionChanged(int position);
    }
}

File: app/src/main/java/com/juniperphoton/myerlist/widget/SelectCategoryView.kt
package com.juniperphoton.myerlist.widget

import android.content.Context
import android.util.AttributeSet
import android.util.Log
import android.widget.LinearLayout
import com.juniperphoton.myerlist.R
import com.juniperphoton.myerlist.extension.dpToPixel
import com.juniperphoton.myerlist.extension.toResColor
import com.juniperphoton.myerlist.model.ToDoCategory
import com.juniperphoton.myerlist.realm.RealmUtils
import io.realm.Realm

class SelectCategoryView(context: Context, attrs: AttributeSet) : LinearLayout(context, attrs) {
    companion object {
        private val TAG = "SelectCategoryView"
        val CATE_VIEW_SIZE_DP = 24
        val CATE_VIEW_MARGIN_lEFT = 8
    }

    var onSelectionChanged: ((Int) -> Unit)? = null

    var selectedIndex = 0
        set(value) {
            field = value
            updateUi(getChildAt(value) as CateCircleView)
            onSelectionChanged?.invoke(value)
        }

    fun toggleSelection(plus: Boolean) {
        Log.d(TAG, "toggleSelection:" + plus)
        var target = selectedIndex + if (plus) 1 else -1
        if (target >= childCount) target = 0
        if (target < 0) target = childCount - 1
        selectedIndex = target
        onSelectionChanged?.invoke(selectedIndex)
    }

    fun makeViews() {
        removeAllViews()
        val categories = Realm.getDefaultInstance().where(ToDoCategory::class.java).findAll()
        categories.filterIndexed handle@ { index, _ ->
            return@handle index >= 0
        }.forEach {
            val circleView = CateCircleView(context, null)
            val layoutParams = LinearLayout.LayoutParams(context.dpToPixel(CATE_VIEW_SIZE_DP), context.dpToPixel(CATE_VIEW_SIZE_DP))
            layoutParams.setMargins(context.dpToPixel(CATE_VIEW_MARGIN_lEFT), 0, 0, 0)
            circleView.layoutParams = layoutParams
            circleView.color = it.intColor
            circleView.setOnClickListener {
                selectedIndex = indexOfChild(it)
            }
            addView(circleView)
        }

        val circleView = CateCircleView(context, null)
        val layoutParams = LinearLayout.LayoutParams(context.dpToPixel(CATE_VIEW_SIZE_DP), context.dpToPixel(CATE_VIEW_SIZE_DP))
        circleView.layoutParams = layoutParams
        circleView.color = R.color.MyerListBlue.toResColor()
        circleView.setOnClickListener {
            selectedIndex = indexOfChild(it)
        }
        addView(circleView, 0)
        selectedIndex = 0
    }

    private fun updateUi(circleView: CateCircleView) {
        (0..childCount - 1)
                .map { getChildAt(it) as CateCircleView }
                .forEach {
                    it.inSelected = it === circleView
                }
    }
}
